feat(core): new Convar txAdmin-suppressEvents

araynimax opened the pull request:
Implemented the feature-request of #595
This also fixes a missing event trigger of
'txAdmin:events:announcement' from
announcements created in the game menu

araynimax committed:
feat(core): new Convar txAdmin-supressEvents
Implemented the feature-request of #595
This also fixes a missing event trigger of
'txAdmin:events:announcement' from
announcements created in the game menu

TasoOneAsia commented on resource/menu/shared.lua:
```diff
@@ -26,6 +26,20 @@ function txPrint(...)
   print(msg)
 end
 
+local supressedEvents = GetConvar('txAdmin-supressEvents', '')
```
Small typo
```suggestion
local supressedEvents = GetConvar('txAdmin-suppressEvents', '')
```

araynimax replied:
👍

TasoOneAsia commented on resource/menu/shared.lua:
```diff
@@ -26,6 +26,20 @@ function txPrint(...)
+  end
+
+  return false
+end
```
It is possible to slightly reduce time complexity by splitting and transforming the ConVar string into a map table. Then, once in the function, compare eventName against the string index in the map table. Reduces time complexity at the cost of space complexity.

If what I said makes no sense, don't worry about it, and keep it as is. This isn't a huge deal at all, and actual real-world use cases will see minimal difference. Change is up to you.

araynimax replied:
Sounds good. I will adjust it that way.

TasoOneAsia reviewed (commented):
Thanks for the PR! Left a couple preliminary comments before @tabarra checks it out. I'd also like to see some documentation for this new ConVar. As it stands, we don't have a central place that documents all of the available txAdmin convars. The menu convars are located in the `docs/menu.md` file, but it may be beneficial to consolidate all of the convars into its own `ConVar.md` file or something of that sort. I'll let @tabarra comment on that direction.

araynimax committed:
fix(core): relocated method
Changed location of the function
since it did not make sense to have it inside the shared.lua

araynimax committed:
feat(doc): added documentation

araynimax commented:
I added the convar documentation into the readme.md and docs/events.md. I think that are good places to find it as a dev.

tabarra commented:
Hey @araynimax, thanks for the contribution!  
  
From my post a few months ago I changed my mind about suppressing/bypassing events.  
If you look at the `feat/core-playerlist` you will see that right now I'm doing a more `txAdmin:event`-centric approach to tx > fxserver communications.  
I even removed a few commands in favor of events:  
![image](https://user-images.githubusercontent.com/1808295/199562371-933a5d4b-d3c0-4d08-964e-b06ebdf9bbe3.png)
  
So i do not want the events to be bypassed, but this still brings the question, how do we handle people wanting to handle announcements/dm messages? I don't know, but that's much more a "ugly notistack bypass" than an event bypass per se.  
Also, considering we only have announcements and DM to overwrite, this could totally be a checkbox in the settings.  
  
That being said, I'll close this PR now and we can discuss how to better handle those cases back in #595 .
